packages/node/src/indexer/fetch.service.ts

```diff
@@ -259,108 +228,146 @@ export class FetchService implements OnApplicationShutdown {
-            startBlockHeight,
-            queryEndBlock,
-            this.nodeConfig.batchSize,
-            this.dictionaryQueryEntries,
-          );
-          //TODO
-          // const specVersionMap = dictionary.specVersions;
-          if (
-            dictionary &&
-            this.dictionaryValidation(dictionary, startBlockHeight)
-          ) {
-            const { batchBlocks } = dictionary;
-            if (batchBlocks.length === 0) {
-              this.setLatestBufferedHeight(
-                Math.min(
-                  queryEndBlock - 1,
-                  dictionary._metadata.lastProcessedHeight,
-                ),
-              );
-            } else {
-              this.blockNumberBuffer.putAll(batchBlocks);
-              this.setLatestBufferedHeight(batchBlocks[batchBlocks.length - 1]);
-            }
-            this.eventEmitter.emit(IndexerEvent.BlocknumberQueueSize, {
-              value: this.blockNumberBuffer.size,
-            });
-            continue; // skip nextBlockRange() way
+          if (this.isShutdown) {
+            resolve();
+            return;
           }
-          // else use this.nextBlockRange()
+          if (this.blockBuffer.freeSpace < this.nodeConfig.batchSize) return;
+          if (isFetchingBlocks) return;
+
+          isFetchingBlocks = true;
+
+          await this.queueBlocks(processor);
+
+          isFetchingBlocks = false;
         } catch (e) {
-          logger.debug(`Fetch dictionary stopped: ${e.message}`);
-          this.eventEmitter.emit(IndexerEvent.SkipDictionary);
+          logger.error(e, `Failed to enqueue blocks for processing`);
+          // TODO should blockBufferSubscription be cleaned up
+          reject(e);
+          sub();
         }
-      }
-      // the original method: fill next batch size of blocks
-      const endHeight = this.nextEndBlockHeight(startBlockHeight);
-      this.blockNumberBuffer.putAll(range(startBlockHeight, endHeight + 1));
-      this.setLatestBufferedHeight(endHeight);
-    }
-  }
+      });
 
-  async fillBlockBuffer(): Promise<void> {
-    while (!this.isShutdown) {
-      const takeCount = Math.min(
-        this.blockBuffer.freeSize,
-        this.nodeConfig.batchSize,
-      );
+      this.blockBufferSubscription = () => {
+        sub();
+        resolve();
+      };
+    });
 
-      if (this.blockNumberBuffer.size === 0 || takeCount === 0) {
-        await delay(1);
-        continue;
-      }
+    // Load initial blocks
+    await this.queueBlocks(processor);
+
+    return task;
+  }
+
+  private async queueBlocks(
+    processor: (value: BlockContent) => Promise<void> | void,
+  ): Promise<void> {
+    let bufferBlocks = await this.nextBlocks();
 
-      const bufferBlocks = await this.blockNumberBuffer.takeAll(takeCount);
-      const metadataChanged = await this.fetchMeta(
-        bufferBlocks[bufferBlocks.length - 1],
+    if (!bufferBlocks.length) {
+      const wait = 3;
```

Comment:
I'm not sure if we want this to be configurable or wait batch size when up to date with the chain.